Add tests for the block synchronizer forward sync flow

alsrdn opened the pull request:
Fixes: https://github.com/casper-network/casper-node/issues/3777

alsrdn committed:
testing: add builder for creating test blocks

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer: add forward sync flow tests

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
testing: implement random distribution for SecretKey

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer/tests: remove unneeded checks

Removes some unneeded checks and steps from a couple of tests.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented:
I'll wait with the review until the fix for `having_block_body_for_block_without_deploys_requires_only_signatures` lands.

alsrdn committed:
block_synchronizer/tests: temporarily ignore failing tests

Ignore tests that are failing until later commits where we can fix the
block_synchronizer behavior.
Also adjust the naming of some tests.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Fraser999 commented on types/src/crypto/asymmetric_key.rs:
```diff
@@ -443,6 +443,21 @@ impl SecretKey {
     }
 }
 
+#[cfg(any(feature = "testing", test))]
+impl rand::distributions::Distribution<SecretKey> for rand::distributions::Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> SecretKey {
+        if rng.gen() {
+            let mut bytes = [0u8; SecretKey::ED25519_LENGTH];
+            rng.fill_bytes(&mut bytes[..]);
+            SecretKey::ed25519_from_bytes(bytes).unwrap()
+        } else {
+            let mut bytes = [0u8; SecretKey::SECP256K1_LENGTH];
+            rng.fill_bytes(&mut bytes[..]);
+            SecretKey::secp256k1_from_bytes(bytes).unwrap()
+        }
+    }
+}
+
```
I'd prefer if we didn't add this.  We've had to add some impls like this to support e.g. proptest tests, but I prefer to stick to the convention of providing a `random` function which takes a `TestRng` specifically, to try to avoid usage of other rngs which don't dump a seed on failure, making test failures harder to reproduce.

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3816/commits/2b0eefcbeb36d4742f2b0fc35b4e36711d4a5f25

Fraser999 commented on node/src/testing/test_block_builder.rs:
```diff
@@ -0,0 +1,194 @@
+
+const MAX_ERA_FOR_RANDOM_BLOCK: u64 = 6;
+
+pub(crate) struct TestBlockBuilder {
```
I'd have probably put this in a `block_builder` module inside the `block` module, just to avoid this `testing` module becoming a dumping ground for all test builder helpers.  Not a big deal though.

alsrdn replied:
Moved in https://github.com/casper-network/casper-node/pull/3816/commits/2b0eefcbeb36d4742f2b0fc35b4e36711d4a5f25

Fraser999 commented on node/src/testing/test_block_builder.rs:
```diff
@@ -0,0 +1,194 @@
+const MAX_ERA_FOR_RANDOM_BLOCK: u64 = 6;
+
+pub(crate) struct TestBlockBuilder {
+    era: Option<u64>,
+    height: Option<u64>,
+    protocol_version: ProtocolVersion,
+    is_switch: Option<bool>,
+    deploys: Vec<Deploy>,
+    state_root_hash: Option<Digest>,
+    parent_hash: Option<BlockHash>,
+    timestamp: Option<Timestamp>,
```
Would be nice to order these in the same way as inside `Block` (and also for the functions below).

alsrdn replied:
Reordered in https://github.com/casper-network/casper-node/pull/3816/commits/2b0eefcbeb36d4742f2b0fc35b4e36711d4a5f25

Fraser999 commented on node/src/testing/test_block_builder.rs:
```diff
@@ -0,0 +1,194 @@
+        self
+    }
+
+    pub(crate) fn build<R: Rng + ?Sized>(self, rng: &mut R) -> Block {
```
Again, I'd prefer if this took a `TestRng` rather than `R: Rng`.

alsrdn replied:
Changed in https://github.com/casper-network/casper-node/pull/3816/commits/2b0eefcbeb36d4742f2b0fc35b4e36711d4a5f25

Fraser999 reviewed (changes requested):
Initial review - I'll complete my review asap.

Blocking due to the use of `R:Rng`, but I also dislike having two ways of doing the same thing.  I think we should create a ticket to remove all the existing `Block::random_xxx` functions, using the new `TestBlockBuilder` instead, and to maybe leave only `Block::random`, but have it implemented using a `TestBlockBuilder`.

alsrdn committed:
block_synchronizer/tests: use TestRng for creating the TestEnv

Force use of `TestRng` consistently within tests.
Also move `TestBlockBuilder` to a module within `block` and re-arrange
fields to better resemble `Block`.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
block_synchronizer/tests: fix formatting

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn commented:
> Initial review - I'll complete my review asap.
> 
> Blocking due to the use of `R:Rng`, but I also dislike having two ways of doing the same thing. I think we should create a ticket to remove all the existing `Block::random_xxx` functions, using the new `TestBlockBuilder` instead, and to maybe leave only `Block::random`, but have it implemented using a `TestBlockBuilder`.

I also dislike having two ways of doing things but I also dislike the amount of boilerplate and code duplication that we have today with `Block::random_with_*_and_*_*` type functions. It's a pain to extend that functionality when we want to add some other configurable parameter to use for block generation. I agree that we should create a new ticket to replace the other uses in the tests with this builder.

alsrdn committed:
block/test_block_builder: fix lint errors

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Fraser999 commented on node/src/types/block.rs:
```diff
@@ -2,6 +2,9 @@
 #![allow(clippy::field_reassign_with_default)]
 #![allow(clippy::boxed_local)] // We use boxed locals to pass on event data unchanged.
 
+#[cfg(any(feature = "testing", test))]
```
Should this just be `#[cfg(test)]` until we expose it publicly (or we could change the gating around the re-export in types.rs to be consistent with this).

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3816/commits/8c830a518622c78e4f57c74fa331707740e90648

Fraser999 commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+}
+
+#[tokio::test]
+async fn fwd_sync_gets_peers_only_form_accumulator() {
```
```suggestion
async fn fwd_sync_gets_peers_only_from_accumulator() {
```

alsrdn replied:
Fixed in https://github.com/casper-network/casper-node/pull/3816/commits/8c830a518622c78e4f57c74fa331707740e90648

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -95,35 +180,77 @@ fn check_sync_global_state_event(event: MockReactorEvent, block: &Block) {
+}
+
+impl BlockSynchronizer {
+    // Create an initialized block synchronizer with default config and maximum 5 simultaneous peers
```
A nit (don't hardcode magic number) :)

```suggestion
    // Create an initialized block synchronizer with default config and maximum simultaneous peers
```

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -201,7 +330,7 @@ async fn global_state_sync_wont_stall_with_bad_peers() {
+    check_sync_global_state_event(event, block);
 
     // Wait for the latch to reset
     std::thread::sleep(Duration::from_secs(6));
```
I think I'll request a tech-debt ticket to add an option to inject the latch reset time. Sleeping for 6 seconds in a test is such a waste of time... Wdyt?

alsrdn replied:
We have [this PR](https://github.com/casper-network/casper-node/pull/3810) which makes the latch configurable. Once that's merged we can change the latch interval here so that we don't waste time.

rafal-ch replied:
Great :tada:

Fraser999 commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -95,35 +180,77 @@ fn check_sync_global_state_event(event: MockReactorEvent, block: &Block) {
+    let peers = test_env.peers();
+    let block = test_env.block();
+    let validator_matrix = test_env.gen_validator_matrix();
+    let validators_private_keys = test_env.validator_keys();
```
Can we rename "private keys" to "secret keys" throughout this file please?

(I'm overly nitpicky about this because there is a tendency to name variables `pk` for either public keys or private keys.  If we at least have a convention of favouring "secret key" over "public key" then such an abbreviation is a little less dangerous.)

alsrdn replied:
Changed in https://github.com/casper-network/casper-node/pull/3816/commits/8c830a518622c78e4f57c74fa331707740e90648

Fraser999 reviewed (approved):
A couple of nitpicks, but otherwise LGTM.

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+    let mut block_synchronizer = BlockSynchronizer::new_initialized(&mut rng, validator_matrix);
+
+    // Register block for forward sync
+    assert!(block_synchronizer.register_block_by_hash(*block.hash(), false, true));
```
I know it's not a part of this PR, but can you extend [the comment](https://github.com/casper-network/casper-node/blob/cc95de3a80cdada1d67a84a2178f23bcf10e36ae/node/src/components/block_synchronizer.rs#L275) for `register_block_by_hash` and describe what is the meaning of the returned `bool`?

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+}
+
+#[tokio::test]
+async fn next_action_for_have_weak_finality_is_fetching_block_body() {
```
Judging by just the name, this is very similar to [this guy](https://github.com/casper-network/casper-node/pull/3816/files#diff-ecca1f5c3a37ff0c7cb64100807c2e3074f1dfdd68582dff2dd768f7a313806bR726).

The latter one is more sophisticated, but please check if we don't duplicate some parts of the test unnecessarily.

alsrdn replied:
You're right, these are similar but actually their target is different. This one puts the builder in `HaveWeakFinality` and checks if the next action is fetching the block body. The other test wants to check that fetching the block body only happens after weak finality was reached, and it goes through some steps of registering signatures to see if it doesn't transition too early. I'll rename the test that you referenced to better state its purpose.

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+        block,
+        validators_private_keys
+            .iter()
+            .take(validators_private_keys.len() / 3 + 1),
```
Can you use `let num_sigs_required_for_weak_finality = validators_private_keys.len() / 3` here, as in other tests?

alsrdn replied:
Actually adding one extra sig is correct to account for the lossy integer division. I'll update the test that didn't have this criteria.

rafal-ch commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -336,3 +452,1231 @@ async fn should_not_stall_after_registering_new_era_validator_weights() {
+    );
+}
+
+// TODO: this test fails sometimes depending on the value of the `LegacyRequiredFinality` chainspec
+// parameter. This happens due to an error in determining if signatures are enough to consider that
+// the block has Strict finality if a block is empty. Remove this ignore when the synchronizer
+// behavior is fixed.
```
Good catch, thanks! :+1:

rafal-ch reviewed (approved):
No blockers.

Just few nits, but good-to-go as is :+1:

georgepisaltu commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -69,14 +77,91 @@ impl MockReactor {
+    async fn process_effects(&self, mut effects: Effects<Event>) -> Vec<MockReactorEvent> {
+        let mut events = Vec::new();
+        for effect in effects.drain(0..) {
+            tokio::spawn(async move { effect.await });
```
Should we `await` the `JoinHandle` returned here so that we simulate a sequential execution of the effects in the queue? This would make the process predictable.

alsrdn replied:
We actually can't await here in this format of the `MockReactor` since we don't mock the other node components (Storage, ContractRuntime etc). Effects created through the effect builder create a responder that listens for the result. We don't want to wait for the result to come (because there is no code to respond). We just want a shortcut to assert which event is generated if the other components would have responded.

georgepisaltu replied:
Ah, I remember our offline talk about this now. I understand, disregard what I wrote above.

alsrdn committed:
block_sychronizer/tests: address code review comments

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

georgepisaltu commented on node/src/components/block_synchronizer/tests.rs:
```diff
@@ -95,35 +180,77 @@ fn check_sync_global_state_event(event: MockReactorEvent, block: &Block) {
+    let test_env = TestEnv::random(&mut rng).with_block(
+        TestBlockBuilder::new()
+            .era(1)
+            .deploys([Deploy::random(&mut rng)].iter())
```
Would be useful to add a `random_deploys(count)` method on the builder (I recently used something similar too), but it's not really necessary to do it in this PR.

alsrdn replied:
Added here: https://github.com/casper-network/casper-node/pull/3816/commits/6404ee17f8ffc9e42e42ddf19f905b214791ef02

alsrdn committed:
test_block_builder: add method for creating multiple random deploys

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>
